datafusion-examples/examples/advanced_udaf.rs

```diff
@@ -412,7 +412,7 @@ async fn main() -> Result<()> {
     let df = ctx.table("t").await?;
 
     // perform the aggregation
-    let df = df.aggregate(vec![], vec![geometric_mean.call(vec![col("a")])])?;
+    let df = df.aggregate(vec![], vec![geometric_mean.call(vec![col("a")]).build()])?;
```

Comment:
Sorry for the delay in reviewing this @jayzhan211  -- I have been thinking about this PR and I am struggling

On one hand, having a builder allows for exposing the ability to provide other types of arguments to aggregate functions (e.g. `ORDER BY` or `FILTER`) 

On the other hand, I feel like this API change (requiring a call to `build()` after `geometric_mean.call` makes DataFusion  harder to learn / use as most calls to aggregate functions are made with just the arguments. 

I was thinking about what an ideal  user experience would be. What do you think about something that permits using the existing `expr_fn` API, but also allows easier construction of more advanced aggregate calls

What about something like  this (as today, no need to call build)

```rust
    // geometric_mean(a)
    let agg = geometric_mean.call(vec![col("a")]);
    let df = df.aggregate(vec![], vec![agg])?;
```

And then then to create an `ORDER BY` the user could create and use a builder like this:

```rust
    // geometric_mean(a FILTER b > c)
    let agg = geometric_mean.call(vec![col("a")])
      .agg_builder() // creates a builder that wraps an Expr, no error return
      .filter(col("b").gt(col("c"((
      .build()?; // check here if everything was ok

    let df = df.aggregate(vec![], vec![agg])?;
```


We could probably implement a trait for `Expr` to add `agg_builder` (or maybe just add it directly)

Reply:
It looks good to me to have both style for function call.

How about the API call in expr_fn? Since, udaf is not public function, it is not a free function that could be called anywhere. API call in expr_fn might be used more frequently.

Should we introduce builder mode?

Current state:
expr_fn::first_value(args, distinct, filter, order_by, null_treatment)

Builder mode:
expr_fn::first_value_builder(args).build()

Reply:
I found that we don;t even need `builder()` with trait impl

Reply:
```
// geometric_mean(a FILTER b > c)
    let agg = geometric_mean.call(vec![col("a")])
      .filter(col("b").gt(col("c")))

    let df = df.aggregate(vec![], vec![agg])?;
```